clients/pkg/promtail/targets/file/filetarget.go

```diff
@@ -91,6 +92,7 @@ type FileTarget struct {
 
 	fileEventWatcher   chan fsnotify.Event
 	targetEventHandler chan fileTargetEvent
+	mu                 sync.Mutex
```

Comment:
I want to think a bit more here, is there a race condition somewhere in the filetarget that we're 1) not detecting with the current tests and 2) not fixing because of 1 + not using the mutex (other than in sync)?

Can the later tests with races be fixed by using the mutex more within the filetarget code rather than accessing `target.mu` in the tests themselves?

Reply:
This is the one that I'm least comfortable with.  The issue is primarily dealing with `target.watches` and `target.readers`.

I know that a lot of the filetarget functions are called from within the `sync` code, and those functions reference the `target.readers`.

Alternatively, we could add a getWatchesLen() and getReadersLen() (each doing appropriate mutex things), and update the test code to utilize those

Reply:
Prototyping this locally, I think this can work.  I didn't put any mutex calls in the test code here, but did have to add `getReadersLen()`, `getReader()`, `deleteReader()` functions to support what the test code needed.

I'll do some more experiments tomorrow (and look at it with a clear head), and then push an update if all goes well.